### PR TITLE
Ability to simulate clusters from support bundles

### PR DESCRIPTION
**Pull request by ibrokethecloud**
The PR introduces the ability to simulate a cluster using the support bundle.

This has been verified using support bundles from harvester v1.0.1.

Pre 1.0.1, the packaged support bundle kit skipped some components which could possibly break re-injection of objectsto the apiserver.

The readme has been updated to reflect the additional usage and there is also a quick start guide in the docs.

### CODE REVIEW

**Comment by bk201** [comment]
Nicely done. Not sure why I'm getting this error when simulate a bundle:
```
ERRO[1214] error during creation of resource .16ea06f26e981ba4 with gvr events.k8s.io/v1, Resource=events
ERRO[1214] map[action:sim-generated apiVersion:events.k8s.io/v1 eventTime:2022-04-28T09:57:31.000000Z kind:Event metadata:map[annotations:map[sim.harvesterhci.io/creationTimestamp:2022-04-28T09:57:31Z] creationTimestamp:2022-04-28T09:57:31Z managedFields:[map[apiVersion:v1 fieldsType:FieldsV1 fieldsV1:map[f:count:map[] f:firstTimestamp:map[] f:involvedObject:map[f:apiVersion:map[] f:kind:map[]] f:lastTimestamp:map[] f:message:map[] f:reason:map[] f:source:map[f:component:map[]] f:type:map[]] manager:kube-controller-manager operation:Update time:2022-04-28T09:57:31Z]] name:.16ea06f26e981ba4 namespace:default uid:be85fd86-94f4-40ef-96c1-1b0a34140bef] note:node3_32eada4e-faca-4e91-8644-f6e3fe392213 stopped leading reason:LeaderElection regarding:map[apiVersion:coordination.k8s.io/v1 kind:Lease] reportingController:sim-generated reportingInstance:sim-generated type:Normal]
FATA[1214] error loading namespacedobjects error during creation of unstructured resource Event ".16ea06f26e981ba4" is invalid: metadata.name: Invalid value: ".16ea06f26e981ba4": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```
The support bundle: [supportbundle_2a0ca8c9-4811-4372-8f8b-376a3cc70a94_2022-04-28T10-25-39Z.zip](https://github.com/rancher/support-bundle-kit/files/8622393/supportbundle_2a0ca8c9-4811-4372-8f8b-376a3cc70a94_2022-04-28T10-25-39Z.zip)
The log: [simulator_error.log](https://github.com/rancher/support-bundle-kit/files/8622398/simulator_error.log)

**Comment by weihanglo** [comment]
Awesome! This really saves the lives of engineer! And I love the quickstart guide you wrote.

I am a bit wary of the default directory `~/.sim`. This abbreviation is net uncommon and people might find it cramming data into existing ~/.sim` and breaks stuff. Any change to avoid this or this is really rare so we can leave it as-is?

The other one is whether we should respect XDG directory standards to store data. I know it is overkill for such a feature, just feel like golang itself does this maybe we can do it as well.

**Comment by ibrokethecloud** [comment]
@bk201 the error is caused due to a weirdly named event in the support bundle..

```
- apiVersion: events.k8s.io/v1
  deprecatedCount: 1
  deprecatedFirstTimestamp: "2022-04-28T09:57:31Z"
  deprecatedLastTimestamp: "2022-04-28T09:57:31Z"
  deprecatedSource:
    component: kube-controller-manager
  eventTime: "null"
  kind: Event
  metadata:
    creationTimestamp: "2022-04-28T09:57:31Z"
    managedFields:
    - apiVersion: v1
      fieldsType: FieldsV1
      fieldsV1:
        f:count: {}
        f:firstTimestamp: {}
        f:involvedObject:
          f:apiVersion: {}
          f:kind: {}
        f:lastTimestamp: {}
        f:message: {}
        f:reason: {}
        f:source:
          f:component: {}
        f:type: {}
      manager: kube-controller-manager
      operation: Update
      time: "2022-04-28T09:57:31Z"
    name: .16ea06f26e981ba4
    namespace: default
    resourceVersion: "39158"
    uid: be85fd86-94f4-40ef-96c1-1b0a34140bef
  note: node3_32eada4e-faca-4e91-8644-f6e3fe392213 stopped leading
  reason: LeaderElection
  regarding:
    apiVersion: coordination.k8s.io/v1
    kind: Lease
  type: Normal
```

I am not entirely sure how or why the event was allowed into the API server in the first place. 

I will see if i can write some logic to massage such events if they are going to be uncommon.

**Comment by bk201** [comment]
@ibrokethecloud Thanks, unfortunately, I don't have the cluster anymore. Otherwise, I can check if there is something wrong when generating the bundle.

**Comment by ibrokethecloud** [comment]
@bk201 as discussed, I have introduced a new CRD named `FailedObjects` which keeps a log of all objects which could not be processed, while allowing remaining objects to be processed.

for example, post the change, the attached support bundle gets processed and the `FailedObjects` CR is created

```
(⎈ |default:default)➜  ~ k get failedobjects -o yaml
apiVersion: v1
items:
- apiVersion: supportbundlekit.io/v1
  kind: FailedObject
  metadata:
    creationTimestamp: "2022-05-13T04:18:57Z"
    generation: 1
    name: failedobjects
    resourceVersion: "7194"
    uid: 6e6ca4f2-db7d-4eaf-a0b7-4cd746874ab3
  spec:
  - error: 'Event ".16ea06f26e981ba4" is invalid: metadata.name: Invalid value: ".16ea06f26e981ba4":
      a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters,
      ''-'' or ''.'', and must start and end with an alphanumeric character (e.g.
      ''example.com'', regex used for validation is ''[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'')'
    gvk: events.k8s.io/v1, Kind=Event
    name: .16ea06f26e981ba4
    namespace: default
  - error: 'Event ".16ea06f26e981ba4" is invalid: [reportingController: Required value,
      reportingController: Invalid value: "": name part must be non-empty, reportingController:
      Invalid value: "": name part must consist of alphanumeric characters, ''-'',
      ''_'' or ''.'', and must start and end with an alphanumeric character (e.g.
      ''MyName'',  or ''my.name'',  or ''123-abc'', regex used for validation is ''([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]'')]'
    gvk: /v1, Kind=Event
    name: .16ea06f26e981ba4
    namespace: default
  - error: 'Event "rke2-ingress-nginx-controller-c94ns.16ea0574b95528a3" is invalid:
      message: Invalid value: "": can have at most 1024 characters'
    gvk: events.k8s.io/v1, Kind=Event
    name: rke2-ingress-nginx-controller-c94ns.16ea0574b95528a3
    namespace: kube-system
  - error: 'Event "rke2-ingress-nginx-controller-c94ns.16ea0574b95528a3" is invalid:
      [reportingController: Required value, reportingController: Invalid value: "":
      name part must be non-empty, reportingController: Invalid value: "": name part
      must consist of alphanumeric characters, ''-'', ''_'' or ''.'', and must start
      and end with an alphanumeric character (e.g. ''MyName'',  or ''my.name'',  or
      ''123-abc'', regex used for validation is ''([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]''),
      message: Invalid value: "": can have at most 1024 characters]'
    gvk: /v1, Kind=Event
    name: rke2-ingress-nginx-controller-c94ns.16ea0574b95528a3
    namespace: kube-system
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**Comment by bk201** [comment]
@ibrokethecloud Overall looks good just some more questions (Sorry it's hard to left comments on with Github PR files because there are too many files).
- Any reason to include `rsync` in the Dockerfile.dapper file? I could find a usage.
- For the test fixture bundle folder: `tests/integration/sampleSupportBundle`, could we compress it first or store it somewhere instead?

**Comment by ibrokethecloud** [comment]
@bk101 the `rsync` is needed for the `make generated_files` target in upstream kubernetes to generate the openapi spec I need to compile apiserver within the simulator command. 

```
HEAD is now at 1f86634ff08 Release commit for Kubernetes v1.21.7
+++ [0523 01:34:25] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
find: 'rsync': No such file or directory
find: 'rsync': No such file or directory
Generating prerelease lifecycle code for 27 targets
```

I will see what i can do with the packaging for the sample.

**Review by weihanglo — COMMENTED** [review]
Could we run `test` script in CI pipeline? I guess add a `test` in `ci` command is sufficient.

File: docs/quickstart.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+```
+drwxr-xr-x   5 gaurav gaurav     4096 Apr 19 18:19  supportbundle_101310ea-583b-4191-b33e-d265002214ea_2022-04-19T08-18-38Z
+-rw-rw-r--   1 gaurav gaurav  5828575 Apr 19 18:19  supportbundle_101310ea-583b-4191-b33e-d265002214ea_2022-04-19T08-18-38Z.zip
+```
+
+change directory to the unzipped support bundle directory and run `support-bundle-kit simulator`
+
+```shell
+(⎈ |default:default)➜  supportbundle_101310ea-583b-4191-b33e-d265002214ea_2022-04-19T08-18-38Z support-bundle-kit simulator
```
Comment (weihanglo) [inline docs/quickstart.md:15]: Might be better use just `$` to denote in a shell
```suggestion
$ support-bundle-kit simulator 
```

File: docs/quickstart.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+This can now be used to interact with the cluster.
+
+As part of the supportbundle collection, the bundle contains a ${nodename}.zip file for each node in the cluster in the `$SUPPORTBUNDLE/nodes` directory. 
+
+The simulator will also load the contents of the same into a CRD named `NodeConfig` in `support-bundle-node-info` namespace.
+
+The CRD spec is as follows:
+
+```
+
```
Comment (weihanglo) [inline docs/quickstart.md:37]: nitpick
```suggestion
```go
```

File: go.mod [hunk 3]
```diff
@@ -6,58 +6,83 @@ replace (
...
-	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.20.4
-	k8s.io/kubectl => k8s.io/kubectl v0.20.4
-	k8s.io/kubelet => k8s.io/kubelet v0.20.4
-	k8s.io/kubernetes => k8s.io/kubernetes v1.20.2
-	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.20.4
-	k8s.io/metrics => k8s.io/metrics v0.20.4
-	k8s.io/mount-utils => k8s.io/mount-utils v0.20.4
-	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.20.4
-
+	go.etcd.io/etcd => go.etcd.io/etcd v0.5.0-alpha.5.0.20200910180754-dd1b699fc489 // ae9734ed278b is the SHA for git tag v3.4.13
```
Comment (weihanglo) [inline go.mod:17]: What does this comment refer to?

File: pkg/simulator/objects/apply.go [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	switch obj.GetKind() {
+	//case "Ingress":
+	// Ingress specific housekeeping
+	//	err = ingressCleanup(obj)
```
Comment (weihanglo) [inline pkg/simulator/objects/apply.go:236]: Why are these lines commented?

File: pkg/simulator/objects/apply.go [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+		// ignore errors
+		ns, _ := o.kc.CoreV1().Namespaces().List(o.ctx, metav1.ListOptions{})
+		if ns != nil && len(ns.Items) != 0 {
+			return nil
+		}
+		time.Sleep(5 * time.Second)
+	}
+
+	return fmt.Errorf("timed out waiting for apiserver to be ready")
+}
```
Comment (weihanglo) [inline pkg/simulator/objects/apply.go:360]: Perhaps use `Poll`? https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Poll

File: pkg/simulator/objects/blockdevices_test.go [hunk 6]
```diff
@@ -0,0 +1,1306 @@
...
+      kubernetes.io/hostname: r620-3
+      ndm.harvesterhci.io/device-type: part
+      ndm.harvesterhci.io/parent-device: b2bf49dab59d30c911d1b51b6f464c18
+    managedFields:
+    - apiVersion: harvesterhci.io/v1beta1
+      fieldsType: FieldsV1
+      fieldsV1:
+        f:spec:
+          f:fileSystem:
+            f:forceFormatted: {}
```
Comment (weihanglo) [inline pkg/simulator/objects/blockdevices_test.go:94]: Is it possible to remove all `manageFields` to reduce lines of code?

File: scripts/validate [hunk 7]
```diff
@@ -15,4 +15,4 @@ if [[ -z "$(command -v golangci-lint)" ]]; then
 fi
 
 echo "Running: golangci-lint run"
-golangci-lint run
+golangci-lint run --timeout 5m
```
Comment (weihanglo) [inline scripts/validate:18]: I might miss something, but why we need this timeout?

Comment (ibrokethecloud) [reply scripts/validate:18]: @weihanglo the default timeout is 1m. With the additional code it was taking more than 1 minute, hence the need to bump it up.

File: pkg/simulator/objects/apply.go (hunk 4 above)
Comment (ibrokethecloud) [reply pkg/simulator/objects/apply.go:236]: Removed the lines. Left over from using a newer version of k8s in the initial version, which mean v1beta1 ingress objects needed a bit of housekeeping.

File: pkg/simulator/objects/apply.go (hunk 5 above)
Comment (ibrokethecloud) [reply pkg/simulator/objects/apply.go]: @weihanglo thanks for the suggestion. I have changed the logic to implement `wait.Poll`

File: go.mod (hunk 3 above)
Comment (ibrokethecloud) [reply go.mod]: My bad.. comment from when i was trying to get the go deps sorted for apiserver

File: pkg/simulator/objects/blockdevices_test.go (hunk 6 above)
Comment (ibrokethecloud) [reply pkg/simulator/objects/blockdevices_test.go]: removed.

**Review by weihanglo — APPROVED** [review]
Fantastic! Thank you for your hard work!

**Review by bk201 — APPROVED** [review]
LGTM!

**Pull request merged**